cirq-google/cirq_google/workflow/qubit_placement.py

```diff
@@ -73,3 +81,130 @@ def _json_dict_(self) -> Dict[str, Any]:
 
     def __repr__(self) -> str:
         return _compat.dataclass_repr(self, namespace='cirq_google')
+
+
+def default_topo_node_to_qubit(node: Any) -> cirq.Qid:
+    """The default mapping from `cirq.NamedTopology` nodes and `cirq.Qid`.
+
+    There is a correspondence between nodes and the "abstract" Qids
+    used to construct un-placed circuit. `cirq.get_placements` returns a dictionary
+    mapping from node to Qid. We use this function to transform it into a mapping
+    from "abstract" Qid to device Qid. This function encodes the default behavior used by
+    `RandomDevicePlacer`.
+
+    If nodes are tuples of integers, map to `cirq.GridQubit`. Otherwise, try
+    to map to `cirq.LineQubit` and rely on its validation.
+
+    Args:
+        node: A node from a `cirq.NamedTopology` graph.
+
+    Returns:
+        A `cirq.Qid` appropriate for the node type.
+    """
+
+    try:
+        return cirq.GridQubit(*node)
+    except TypeError:
+        return cirq.LineQubit(node)
```

Comment:
I'd propose to add a method `nodes_to_qubits` in the `NamedTopology` class and use it here instead of this hack, especially because the lambda argument in the qubit placer class below is also not serializable. 

Note that we already have `nodes_to_gridqubits` method on `TiltedSquareLattice` and have an open issue to add a similar `nodes_to_linequbits` method for the `LineTopology`.

Reply:
hmm I don't consider this a hack.

`NamedTopology.nodes_to_qubits` may _seem_ obvious for LineTopology and TiltedSquareLattice but (a) maybe that isn't so and (b) it would be non-obvious for other topology types.

(a) what if you happened to use a different qubit type, especially for LineTopology. You could have written your circuit using GridQubit(x, 0) or even GridQubit(x, y) for fixed y where you know that that defines an actual list on the device. What if you used NamedQubit's a, b, c, ..., 

(b) Imagine additional named topologies that don't have a corresponding qubit type. Consider a hexagonal topology. what would be the nodes_to_qubits function? there are multiple ways to handle this https://www.redblobgames.com/grids/hexagons/ including mapping to gridqubits or defining a new coordinate system.

`nodes_to_gridqubits` and `nodes_to_linequbits` as non-abstract, non-interface, topology-specific methods makes sense. you can even imagine having `LineTopology.nodes_to_gridqubits`. But I think it would be too tricky to have as part of the NamedTopology interface

Reply:
1. I imagine there would be more places where we'll end up needing the `topo_node_to_qubit_func` parameter, i.e. all qubit placement / routing related utilities. It would be nice if we can avoid passing this additional parameter around by putting the mapping in the named topology itself. 
2. Do we imagine a use case of NamedTopology's where the topology itself should exist but the mapping routines need not exist because they are hard to write? IIUC, whenever someone creates a new NamedToplogy (eg: the hexagonal topology), they'd also have to figure out a way to map the topology to qubits for it to be useful. If that's the case, then it's just a matter of whether we should do it outside the class or inside. 
3. Assuming 2 above is true, we can deal with the problem of different mappings for a specific named topology by either deriving from the topology class and overriding the `nodes_to_qubits` method, or by parameterising the topology class with flags to control the placement behavior. 
4. Moving the placement logic to named topology has an added benefit that it makes it serializable. I'm curious to know your perspective on why that is not so important?

Reply:
The true solution would be to let you define `Circuit`s with arbitrary qubit identifiers, and then you could use integers or tuples of integers or whatever is appropriate for your problem. The current construct is a shim around _that_. It's like if in classical programming you couldn't use variable names and had to use hardcoded memory addresses. Most of what follows expands on this idea.

1. any qubit mapping worth its salt shouldn't care about the form of the input "qubit" (logical qubit, node, whatever). The shim is that the input qubits must be `cirq.Qid` even if that doesn't make complete sense. 
2. No; it's not straightforward like that. There are qubit mapping strategies that are completely agnostic to the input and output topologies; some that make make some assumptions (e.g. maybe specialize for a linear or planar target topo); and some basic strategies that make very strict assumptions (like the Naive + offset placer). That's why the mapping should happen in the mapping strategy.
3. Philosophically, this would be a step in the wrong direction. I want to be able to describe \~\~problem\~\~ topologies independent of their implementation. 
4. The placement logic is serializable. If I tell you that I used `NaiveQubitPlacer` to run my problem, you'd know what strategy I used. If I gave you a json document with `RandomQubitPlacer` and the target topology, you'd know what strategy I used.

There's just this annoying thing you need to "integrate out" where I'd like to describe my problem in terms of namedtopology nodes but have to use gridqubit or linequbit

Reply:
some semi-relevant hearsay https://github.com/quantumlib/Cirq/issues/4713